Update Go version to 1.18.4

belimawr opened the pull request:
## What does this PR do?

Update Go version to 1.18.4, update linter config (dependant on Go
version) and fix all lint issues reported.

## Why is it important?

It adds support for Go Workspaces (added on Go 1.18) and brings many security fixes

## Checklist

~~- [ ] My code follows the style guidelines of this project~~
~~- [ ] I have commented my code, particularly in hard-to-understand areas~~
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
~~- [ ] I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.md` or `CHANGELOG-developer.md`.

~~## Author's Checklist~~
~~## How to test this PR locally~~

## Related issues
- Closes #25

~~## Use cases~~
~~## Screenshots~~
~~## Logs~~

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/elastic-agent-inputs-mbp%2FPR-38/detail/PR-38/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/elastic-agent-inputs-mbp%2FPR-38/detail/PR-38/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/elastic-agent-inputs-mbp%2FPR-38/detail/PR-38/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/elastic-agent-inputs-mbp%2FPR-38/detail/PR-38/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://null_38.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-07-26T07:59:12.455Z&rangeTo=2022-07-26T08:19:12.455Z&transactionName=elastic-agent-inputs-mbp/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=58594593754db77a725a8ccec05610e1&transactionId=4d58ec1f7daaad47)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-07-26T08:09:12.455+0000


* Duration: 2 min 27 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![3](https://img.shields.io/badge/3%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/elastic-agent-inputs-mbp%2FPR-38/detail/PR-38/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Check out from version control`
<ul>
<li>Took 0 min 10 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/elastic-agent-inputs-mbp/pipelines/PR-38/runs/1/steps/10/log/?start=0">here</a></li>
<li>Description: <code>[2022-07-26T08:10:48.578Z] The recommended git tool is: NONE
[2022-07-26T08:10:56.683Z] using crede</code></l1>
</ul>
  
  
  
  
##### `Check out from version control`
<ul>
<li>Took 0 min 2 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/elastic-agent-inputs-mbp/pipelines/PR-38/runs/1/steps/14/log/?start=0">here</a></li>
<li>Description: <code>[2022-07-26T08:11:09.121Z] The recommended git tool is: NONE
[2022-07-26T08:11:09.132Z] using crede</code></l1>
</ul>
  
  
  
  
##### `Check out from version control`
<ul>
<li>Took 0 min 2 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/elastic-agent-inputs-mbp/pipelines/PR-38/runs/1/steps/18/log/?start=0">here</a></li>
<li>Description: <code>[2022-07-26T08:11:25.629Z] The recommended git tool is: NONE
[2022-07-26T08:11:25.649Z] using crede</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

cmacknz commented on .golangci.yml:
```diff
@@ -4,17 +4,23 @@
 run:
   # timeout for analysis, e.g. 30s, 5m, default is 1m
-  timeout: 1m
+  timeout: 15m
```
I hope lint doesn't usually take 15 minutes now

belimawr replied:
The mage target uses a template from `dev-tools/templates/.golangci.yml` that differs than the one from the project root. I kept the template one because it was more similar to the Beats one.

But now I'm wondering if it was a good decision... I'll review the template and the overall changes.

cmacknz replied:
Ah fair enough. I'm going to remove the templates everywhere, it was a convenience to update the Go version which is now automated.

cmacknz commented on .golangci.yml:
```diff
@@ -4,17 +4,23 @@
+  exclude-rules:
+    # Exclude package name contains '-' issue because we have at least one package with
+    # it on its name.
+    - text: "ST1003:"
```
You could just fix this one in this repository, unless you specifically want to keep allowing it

belimawr replied:
I'll fix it ;)

As I mentioned on the other comment, I didn't pay much attention to the contents, just made sure the linter was working.

cmacknz commented on .golangci.yml:
```diff
@@ -129,38 +145,22 @@ linters-settings:
-            reason: "This package is deprecated"
-        - github.com/elastic/beats/v7:
-            reason: "There must be no Beats dependency"
+    require-specific: false
```
Does this just allow writing `//nolint` instead of `//nolint: staticcheck`? I like mentioning the specific linter but I don't care that much about it.

belimawr replied:
Indeed, I'll change it to true.

cmacknz commented on .github/workflows/golangci-lint.yml:
```diff
@@ -26,24 +26,24 @@ jobs:
         run: echo Go GOOS=$GOOS
 
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v3
```
What does this fix?

cmacknz replied:
Ah just a general version bump? https://github.com/golangci/golangci-lint-action/releases

belimawr replied:
Yes, I was fighting with an issue that only happened on CI, so I ended up updating golangci-lint-action action and its "dependencies".

At the end it was all my fault because my PR didn't have the latest commit from `main`, hence I didn't get any lint errors locally.

cmacknz replied:
Let's keep the update anyway

cmacknz commented on .github/workflows/golangci-lint.yml:
```diff
@@ -26,24 +26,24 @@ jobs:
         with:
           # Optional: version of golangci-lint to use in form of v1.2 or v1.2.3 or `latest` to use the latest version
-          version: v1.44.2
+          version: v1.45.2
```
You may as well bump to the latest 1.47.2 version, they have been slowly adding support for Go 1.18

cmacknz replied:
I will do this in our other repositories.

belimawr commented:
@cmacknz I implemented all suggestions and also added `io/ioutil` as a package that cannot be imported (it has been deprecated since Go 1.16). The changes were minimal.

cmacknz commented on dev-tools/templates/.golangci.yml:
```diff
@@ -114,6 +114,12 @@ linters-settings:
+    # Check the list against standard lib.
+    include-go-root: true
+    packages-with-error-message:
+      - io/ioutil: "The package is deprecated, use `io` or `so` instead."
```
```suggestion
      - io/ioutil: "The package is deprecated, use `io` or `os` instead."
```
